heareval/baseline.py

```diff
@@ -121,25 +120,30 @@ def frame_audio(audio: Tensor, frame_size: int, hop_size: int, is_centered: bool
     """
     batch_size, num_samples = audio.shape
 
-    # Adjust the number of samples if centering to allow for a half
-    # frame size number of padded samples padded at the start.
+    # If centering, we are going to pad the start and end of the clip
+    # by frame_size // 2. Here, we increase the number of samples to that
+    # size before calculating the number of frames for this hop size.
     start_pad = 0
     if is_centered:
-        half_frame_size = int(frame_size // 2)
-        num_samples += half_frame_size
-        start_pad = half_frame_size
+        start_pad = int(frame_size // 2)
+        num_samples += start_pad * 2
+
+    # Number of frames that will fully fit within num_samples.
+    num_frames = 1 + (num_samples - frame_size) // hop_size
 
-    # Number of frames is the number of hops that can occur within num_samples
-    num_frames = math.ceil(num_samples / hop_size)
+    # Number of samples after applying framing.
+    framed_num_samples = (num_frames - 1) * hop_size + frame_size
 
-    # Pad audio to facilitate centered frames.
-    padded_num_samples = (num_frames - 1) * hop_size + frame_size
-    end_pad = padded_num_samples - num_samples
-    audio = F.pad(audio, (start_pad, end_pad))
+    # If we are centering, then we will need to pad the audio, if not
+    # centering, then we will possibly be truncating the input audio.
+    if is_centered:
+        end_pad = framed_num_samples - audio.shape[1] - start_pad
+        audio = F.pad(audio, (start_pad, end_pad))
+    else:
+        audio = audio[:, :framed_num_samples]
 
-    # Frame.
     shape = (batch_size, num_frames, frame_size)
-    stride = (1, hop_size, 1)
+    stride = (audio.stride()[0], hop_size, 1)
```

Comment:
This was the main bug, it was a bit of a mind boggle for me lol. But, audio.stride()[0] tells us how much we need to stride in order to use this in batches. For example take this tensor of shape (3,10).

```
x = torch.arange(30).view(3,10)

tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24, 25, 26, 27, 28, 29]])
```

And we want a frame_rate of 5 and a frame_size of 5. 

What we were doing before is this:
```
x.as_strided(size=(3, 2, 5), stride=(1, 5, 1))

tensor([[[ 0,  1,  2,  3,  4],
         [ 5,  6,  7,  8,  9]],

        [[ 1,  2,  3,  4,  5],
         [ 6,  7,  8,  9, 10]],

        [[ 2,  3,  4,  5,  6],
         [ 7,  8,  9, 10, 11]]])
```

But what we want is this:
```
x.as_strided(size=(3, 2, 5), stride=(x.stride()[0], 5, 1))

tensor([[[ 0,  1,  2,  3,  4],
         [ 5,  6,  7,  8,  9]],

        [[10, 11, 12, 13, 14],
         [15, 16, 17, 18, 19]],

        [[20, 21, 22, 23, 24],
         [25, 26, 27, 28, 29]]])

x.stride()[0] == 10
```

Reply:
The batches were not correctly being synthesized and thus the reason for the failure of the other test case of the models as well. This is a great find.